### PR TITLE
Add discard_classes option to support strings & regular expressions and deprecate ignore_classes

### LINKED ISSUE #589: config.ignore_classes should deal in strings
### Description
`config.ignore_classes` should take an array of strings instead of constants (or either?)

### Issue
As we add error classes to the `ignore_classes` we're realizing that we're inadvertently loading those classes that don't need to be loaded. It would be good if `ignore_classes` can take in an array of class names as strings since this comparison is done at the time of error anyways and the raised error can just be converted to a string for comparison.

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

The `ignore_classes` option currently supports `Class` constants and `Proc`. Rails has recently deprecated autoloading in initialization (where Bugsnag is usually configured) which makes `Class` constants unusable. Support for `Proc` is powerful but un-ergonomic when simply trying to ignore an error by class name

In order to help bring the Ruby notifier in-line with other notifiers I've added `discard_classes` as a new option and deprecated the existing `ignore_classes` option. `discard_classes` works similarly to `ignore_classes`, with these differences:

1. It accepts `String` or `Regexp` rather than `Class` or `Proc`
2. It doesn't walk the ancestor chain

### Linked issues

Fixes #589

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: lib/bugsnag/middleware/ignore_error_class.rb [hunk 1]
```diff
@@ -10,9 +10,16 @@ def initialize(bugsnag)
     def call(report)
       ignore_error_class = report.raw_exceptions.any? do |ex|
         ancestor_chain = ex.class.ancestors.select { |ancestor| ancestor.is_a?(Class) }.to_set
+        ancestor_chain_strings = ancestor_chain.map(&:to_s)
```
Comment (tomlongridge) [inline lib/bugsnag/middleware/ignore_error_class.rb:13]: Having seen this in the test, I wonder whether matching strings and regexes against ancestors is confusing. I can see if you pass an actual class in then matching subclasses sounds reasonable, but with strings and particularly regexes I think we should probably just match against the exception class. But I guess it could be argued that it would be confusing to have two different behaviours...

Comment (imjoehaines) [reply lib/bugsnag/middleware/ignore_error_class.rb:13]: Yeah I can see the arguments both ways

I went for this way because the migration path to fix the Rails deprecation is as easy as wrapping the current values in quotes

If we do an exact match then you'd need to find out each concrete exception class you want to ignore if you were relying on ignoring an ancestor using a class constant

Then because of that, `Regexp` being the only value handled differently would feel weird, though I can see how it might be very unexpected to match against every ancestor

File: lib/bugsnag/middleware/discard_error_class.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    ##
+    # @param middleware [#call] The next middleware to call
+    def initialize(middleware)
+      @middleware = middleware
+    end
+
+    ##
+    # @param report [Report]
+    def call(report)
+      ignore_error_class = report.raw_exceptions.any? do |ex|
```
Comment (tomlongridge) [inline lib/bugsnag/middleware/discard_error_class.rb:15]: I know it's just a local variable name, but perhaps better refer to it as "discarding" rather than "ignoring" here?

File: lib/bugsnag/configuration.rb [hunk 3]
```diff
@@ -123,6 +128,7 @@ def initialize
       # SystemExit and SignalException are common Exception types seen with
       # successful exits and are not automatically reported to Bugsnag
       self.ignore_classes = Set.new([SystemExit, SignalException])
+      self.discard_classes = Set.new(["SystemExit", "SignalException"])
```
[the remark this answers is not in the document]
Comment (imjoehaines) [reply lib/bugsnag/configuration.rb:131]: Yes, so long as we remember to move them :p

How do we document it in that case? Currently we [list the default values in `ignore_classes`](https://docs.bugsnag.com/platforms/ruby/other/configuration-options/#ignore_classes) but if we add a deprecation notice and point people to `discard_classes`, I think we'd want to list the defaults there

Alternatively would you be ok with building the `discard_classes` from `ignore_classes` to remove the repetition? i.e. something like:

```ruby
self.ignore_classes = Set.new([SystemExit, SignalException])
self.discard_classes = self.ignore_classes.map(:&to_s)
```

[the remark this answers is not in the document]
Comment (imjoehaines) [reply lib/bugsnag/configuration.rb]: Removed the defaults in 4cced91b54b891f598b8a35d1196c439508a11ad

**Comment by pic** [comment]
support of `Proc` would still be useful so in the proc I could check `is_a?` and still be able to group the exceptions that I want to ignore with a common ancestor. Is there any chance that it will be added again?

**Comment by snmaynard** [comment]
You should be able to use the newly added `on_error` Procs to handle that

**Comment by ngan** [comment]
So is the suggested way to use `on_error` for walking ancestors? Or continue using adding a `Proc` to `ignore_classes`?

**Comment by imjoehaines** [comment]
> So is the suggested way to use `on_error` for walking ancestors? Or continue using adding a `Proc` to `ignore_classes`?

Yeah I'd recommend using `on_error` for that. If you're already using `ignore_classes` then it will continue to work for now, but you'll need to migrate eventually
